domain/secretbackend/service/modelsecretbackendservice.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/juju/errors"
+
+	coremodel "github.com/juju/juju/core/model"
+	secretbackenderrors "github.com/juju/juju/domain/secretbackend/errors"
+	"github.com/juju/juju/internal/secrets/provider"
+	"github.com/juju/juju/internal/secrets/provider/kubernetes"
+)
+
+// ModelSecretBackendService is a service for interacting with the secret backend state for a specific model.
+type ModelSecretBackendService struct {
+	st      State
+	modelID coremodel.UUID
+}
+
+// NewModelSecretBackendService creates a new ModelSecretBackendService for interacting with the secret backend state for a specific model.
+func NewModelSecretBackendService(modelID coremodel.UUID, st State) *ModelSecretBackendService {
+	return &ModelSecretBackendService{modelID: modelID, st: st}
+}
+
+// GetModelSecretBackend returns the secret backend name for the current model ID,
+// returning an error satisfying [modelerrors.NotFound] if the model provided does not exist.
+func (s *ModelSecretBackendService) GetModelSecretBackend(ctx context.Context) (string, error) {
+	modelSecretBackend, err := s.st.GetModelSecretBackendDetails(ctx, s.modelID)
+	if err != nil {
+		return "", fmt.Errorf("cannot get model secret backend detail for %q: %w", s.modelID, err)
+	}
+	backendName := modelSecretBackend.SecretBackendName
+	switch modelSecretBackend.ModelType {
+	case coremodel.IAAS:
+		if backendName == provider.Internal {
+			backendName = provider.Auto
+		}
+	case coremodel.CAAS:
+		if backendName == kubernetes.BackendName {
+			backendName = provider.Auto
+		}
+	default:
+		// Should never happen.
```

Comment:
I understand that this is logic that we already have in the system and we are just moving it over here. But this logic just seems broken.

We have the backend name and if we don't understand what the model type is we can just return that and avoid an error. This is a very contrived situation we have made all to do some weird stuff with auto.

Not a show stopper from this PR but someone has to ask the question does this actually make any sense?